packages/angular/ssr/src/routes/ng-routes.ts

```diff
@@ -0,0 +1,183 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { APP_BASE_HREF, PlatformLocation } from '@angular/common';
+import {
+  ApplicationRef,
+  Compiler,
+  Injector,
+  createPlatformFactory,
+  platformCore,
+  ɵwhenStable as whenStable,
+  ɵConsole,
+} from '@angular/core';
+import {
+  INITIAL_CONFIG,
+  ɵINTERNAL_SERVER_PLATFORM_PROVIDERS as INTERNAL_SERVER_PLATFORM_PROVIDERS,
+} from '@angular/platform-server';
+import { Route, Router, ɵloadChildren as loadChildrenHelper } from '@angular/router';
+import { Console } from '../console';
+import { AngularAppManifest } from '../manifest';
+import { isNgModule } from '../utils/ng';
+import { joinUrlParts } from '../utils/url';
+
+/**
+ * Represents the result of processing a route.
+ */
+export interface RouteResult {
+  /**
+   * The resolved path of the route.
+   *
+   * This string represents the complete URL path for the route after it has been
+   * resolved, including any parent routes or path segments that have been joined.
+   */
+  route: string;
+
+  /**
+   * The target path for route redirection, if applicable.
+   *
+   * If this route has a `redirectTo` property in the configuration, this field will
+   * contain the full resolved URL path that the route should redirect to.
+   */
+  redirectTo?: string;
+}
+
+/**
+ * Retrieves routes from the given Angular application.
+ *
+ * Initializes an Angular platform, bootstraps the application or module,
+ * and retrieves routes from the Angular router configuration. Handles both module-based and
+ * function-based bootstrapping. Yields the resulting routes as `RouteResult` objects.
+ *
+ * @param bootstrap - A function that returns a promise resolving to an `ApplicationRef` or an Angular module to bootstrap.
+ * @param document - The initial HTML document used for server-side rendering.
+ * @param url - The URL for server-side rendering.
+ * @returns An async iterator yielding `RouteResult` objects.
+ */
+export async function* getRoutesFromAngularRouterConfig(
```

Comment:
Do we use a similar logic when extracting routes for SSG purposes? If that's the case, I'm wondering if we can eventually share the logic to avoid duplication and simplify maintenance.

Reply:
Yes, that is a goal.